src/main/java/net/ripe/rpki/rsyncit/rsync/RsyncWriter.java

```diff
@@ -45,84 +52,90 @@ public RsyncWriter(Config config) {
+                var writableContent = filterOutBadUrls(hostDirectory, os).stream()
+                        .map(rpkiObject -> {
+                            var relativeUriPath = hostUrl.relativize(rpkiObject.url()).getPath();
+                            var targetPath = hostDirectory.resolve(relativeUriPath).normalize();
+
+                            assert targetPath.normalize().startsWith(hostDirectory.normalize());
+
+                            return new ObjectTarget(targetPath, rpkiObject.bytes(), FileTime.from(rpkiObject.modificationTime()));
+                        }).toList();
+
+                // Create directories
+                // Since createDirectories is idempotent, we do not worry about the order in which it is actually
+                // executed. However, we do want a stable sort for .distinct()
+                var targetDirectories = writableContent.stream().map(o -> o.targetPath.getParent())
+                        .sorted(Comparator.comparing(Path::getNameCount).thenComparing(Path::toString))
+                        .distinct().toList();
+
+                var t0 = System.currentTimeMillis();
+                fileWriterPool.submit(() -> targetDirectories.parallelStream()
+                        .forEach(dir -> {
+                            try {
+                                Files.createDirectories(dir);
+                                Files.setPosixFilePermissions(dir, DIRECTORY_PERMISSIONS);
+                                Files.setLastModifiedTime(dir, INTERNAL_DIRECTORY_LAST_MODIFIED_TIME);
+                            } catch (IOException e) {
+                                log.error("Could not create directory {}", dir, e);
+                                throw new UncheckedIOException(e);
+                            }
+                        })
                 ).join();
+
+                var t1 = System.currentTimeMillis();
+                fileWriterPool.submit(() -> writableContent.parallelStream().forEach(content -> {
+                    try {
+                        Files.write(content.targetPath, content.content);
+                        Files.setPosixFilePermissions(content.targetPath, FILE_PERMISSIONS);
+                        Files.setLastModifiedTime(content.targetPath, content.modificationTime);
+                    } catch (IOException e) {
+                        throw new UncheckedIOException(e);
+                    }
+                })).join();
+
+                log.info("Wrote {} directories ({} ms) and {} files ({} ms) for host {}",
+                        targetDirectories.size(), t1 - t0,
+                        writableContent.size(), System.currentTimeMillis() - t1,
+                        hostName);
             });
 
+            // Calculate target directory after writing phase, to be sure it is not used beforehand.
+            final Path targetDirectory = generatePublicationDirectoryPath(config.rsyncPath(), now);
```

Comment:
Since we pass the timestamp and `generatePublicationDirectoryPath` does not check whether or not the target exists, I don't think the comment is true :)